pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -461,28 +466,38 @@ func gcListWatcher(client *dynamic.Client, resource unversioned.GroupVersionReso
 	}
 }
 
-func monitorFor(p *Propagator, clientPool dynamic.ClientPool, resource unversioned.GroupVersionResource) (monitor, error) {
+func monitorFor(p *Propagator, clientPool dynamic.ClientPool, resource unversioned.GroupVersionResource, kind unversioned.GroupVersionKind) (monitor, error) {
 	// TODO: consider store in one storage.
 	glog.V(6).Infof("create storage for resource %s", resource)
 	var monitor monitor
-	client, err := clientPool.ClientForGroupVersion(resource.GroupVersion())
+	client, err := p.gc.metaOnlyClientPool.ClientForGroupVersion(resource.GroupVersion())
 	if err != nil {
 		return monitor, err
 	}
+	setObjectTypeMeta := func(obj interface{}) {
+		runtimeObject, ok := obj.(runtime.Object)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("expected runtime.Object, got %#v", obj))
+		}
+		runtimeObject.GetObjectKind().SetGroupVersionKind(kind)
+	}
 	monitor.store, monitor.controller = framework.NewInformer(
 		gcListWatcher(client, resource),
 		nil,
 		ResourceResyncTime,
 		framework.ResourceEventHandlerFuncs{
 			// add the event to the propagator's eventQueue.
 			AddFunc: func(obj interface{}) {
+				setObjectTypeMeta(obj)
 				event := event{
 					eventType: addEvent,
 					obj:       obj,
 				}
 				p.eventQueue.Add(event)
 			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
+				setObjectTypeMeta(newObj)
```

Comment:
Chao convinced me IRL :(

On Thu, Aug 4, 2016 at 2:26 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/controller/garbagecollector/garbagecollector.go
> <https://github.com/kubernetes/kubernetes/pull/28480#discussion_r73605893>
> :
>
> >  				event := event{
> >  					eventType: addEvent,
> >  					obj:       obj,
> >  				}
> >  				p.eventQueue.Add(event)
> >  			},
> >  			UpdateFunc: func(oldObj, newObj interface{}) {
> > +				setObjectTypeMeta(newObj)
>
> I don't have a codec. One benefit of registering the MetaOnly to the
> scheme is utilizing the existing codecs, like the DirectCodec. I think it's
> neater to do it here unless we see other use cases then we can put a
> generalized codec.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28480/files/41d92b1b3fe4883311ae3037fb816f858d565fbd#r73605893>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngltTGxnes6BT_iSGGuyLD-HbyUuwZks5qclkWgaJpZM4JEvyT>
> .
>